src/main/oauth2/utils.ts

```diff
@@ -0,0 +1,148 @@
+import { OAUTH_AUTHORIZATION_SERVER_URL, OAUTH_CLIENT_ID, WELL_KNOWN_OAUTH_AUTHORIZATION_SERVER_URL } from "@main/config/server";
+import { generatePKCE } from "@main/oauth2/pkce";
+import RedirectHandler from "@main/oauth2/redirect-handler";
+import { logger } from "@main/utils/logger";
+import { shell } from "electron";
+import { stringify } from "querystring";
+
+const log = logger("oauth2-utils");
+
+interface TokenResponse {
+    token: string;
+    refreshToken: string;
+    expiresIn: number;
+}
+
+export async function fetchAuthorizationServerMetadata(): Promise<{
+    authorizationEndpoint: string;
+    tokenEndpoint: string;
+}> {
+    const response = await fetch(WELL_KNOWN_OAUTH_AUTHORIZATION_SERVER_URL);
+    if (response.status !== 200) {
+        const error = `Failed to fetch OAuth2 authorization server metadata: ${response.status} ${response.statusText}`;
+        log.error(error);
+        throw new Error(error);
+    }
+    const { authorization_endpoint, token_endpoint } = await response.json();
+    if (!authorization_endpoint || !token_endpoint) {
+        const error = "Invalid OAuth2 authorization server metadata";
+        log.error(error);
+        throw new Error(error);
+    }
+    // TODO: Remove this hack once the server is fixed
+    const fixedAuthorizationEndpoint = authorization_endpoint.replaceAll(":8888", "");
+    const fixedTokenEndpoint = token_endpoint.replaceAll(":8888", "");
+    return { authorizationEndpoint: fixedAuthorizationEndpoint, tokenEndpoint: fixedTokenEndpoint };
+}
+
+// Careful with shell.openExternal. https://benjamin-altpeter.de/shell-openexternal-dangers/
+function openInBrowser(url: string) {
+    if (!["https:", "http:"].includes(new URL(url).protocol)) return;
+    // Additional checks to prevent opening arbitrary URLs
+    if (!url.startsWith(OAUTH_AUTHORIZATION_SERVER_URL)) return;
+    shell.openExternal(url);
+}
+
+function createUrlWithQuerystring(baseUrl: string, params: Record<string, string | number | boolean>): string {
```

Comment:
I am not convinced of the value of this function tbh. It's a tiny wrapper around `querystring.stringify` and seems to only add a layer of indirection. The type signature is a mouthful as well.